### PR TITLE
refactor(source): refactor source parser/reader and refine some code

### PR DESCRIPTION
**Pull request by waruto210**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

- introduce new stream based source parser trait `ByteStreamSourceParser` and reader trait `SplitReaderV2`
- move the `parser` mod to `connector`, prepare for refactoring of other parsers and readers
- pushdown the parser to s3 reader
- impl `SplitReaderV2` for DummySplitReader
- refine some code

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
~- [] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

#7032

### CODE REVIEW

File: src/connector/src/parser/csv_parser.rs [hunk 1]
```diff
@@ -44,8 +56,15 @@ pub struct CsvParser {
...
 
 impl CsvParser {
-    pub fn new(delimiter: u8, has_header: bool) -> Result<Self> {
+    pub fn new(rw_columns: Vec<SourceColumnDesc>, parser_config: CsvParserConfig) -> Result<Self> {
+        let CsvParserConfig {
+            delimiter,
+            has_header,
+        } = parser_config;
+
+        // tracing::info!("rw_columns: {:?}", rw_columns);
```
Comment (TennyZhuang) [inline src/connector/src/parser/csv_parser.rs:65]: If it's too verbose, give it another level or just remove the comments.

**Review by TennyZhuang — DISMISSED** [review]

File: src/connector/src/parser/csv_parser.rs (hunk 1 above)
Comment (waruto210) [reply src/connector/src/parser/csv_parser.rs]: removed

**Review by xx01cyx — COMMENTED** [review]
Generally LGTM

**Review by tabVersion — APPROVED** [review]

**Comment by lmatz** [comment]
maybe we merge this after the 0.1.16 release? not sure if any pr with respect to the same issue has been merged

**Comment by waruto210** [comment]
> maybe we merge this after the 0.1.16 release? not sure if any pr with respect to the same issue has been merged

This pr should be the only one that refactor parser/readers.

**Pull request merged**